packages/cloudflare/src/api/kvCache.ts

```diff
@@ -0,0 +1,126 @@
+/* eslint-disable @typescript-eslint/no-unused-vars */
+import type { KVNamespace } from "@cloudflare/workers-types";
+import type { Extension } from "@opennextjs/aws/types/cache";
+import type { CacheValue, IncrementalCache, WithLastModified } from "@opennextjs/aws/types/overrides";
+import { IgnorableError, RecoverableError } from "@opennextjs/aws/utils/error.js";
+
+import { getCloudflareContext } from "./get-cloudflare-context.js";
+
+export const CACHE_ASSET_DIR = "cnd-cgi/_next_cache";
+
+/**
+ * Open Next cache based on cloudflare KV and Assets.
+ *
+ * Note: The class is instantiated outside of the request context.
+ * The cloudflare context and process.env are not initialzed yet
+ * when the constructor is called.
+ */
+class Cache implements IncrementalCache {
+  readonly name = "cloudflare-kv";
+  protected initialized = false;
+  protected kv: KVNamespace | undefined;
+  protected assets: Fetcher | undefined;
+
+  async get<IsFetch extends boolean = false>(
+    key: string,
+    isFetch?: IsFetch
+  ): Promise<WithLastModified<CacheValue<IsFetch>>> {
+    if (!this.initialized) {
+      await this.init();
+    }
+
+    if (!(this.kv || this.assets)) {
+      throw new IgnorableError(`No KVNamespace nor Fetcher`);
+    }
+
+    this.debug(`Get ${key}`);
+
+    try {
+      this.debug(`- From KV`);
+      const kvKey = this.getKVKey(key, isFetch ? "fetch" : "cache");
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      let value: any = await this.kv?.get(kvKey, "json");
+      if (!value && this.assets) {
+        const url = this.getAssetUrl(key);
+        const response = await this.assets.fetch(url);
+        this.debug(`- From Assets`);
+        if (response.ok) {
+          value = await response.json();
```

Comment:
Good question, I need to think about the tradeoff: reducing KV traffic reduces costs

Reply:
Is it not charged for get that doesn't return data ? Because you are already making a request to KV here

Reply:
Write requests are [10x](https://developers.cloudflare.com/kv/platform/pricing/) the price.
But I'll add a comment there to further investigate.
It could be an option if there is a perf boost.